cider/tests/utils/ArrowArrayBuilder.h

```diff
@@ -0,0 +1,158 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+#ifndef CIDER_ARROWARRAYBUILDER_H
+#define CIDER_ARROWARRAYBUILDER_H
+
+#include "cider/CiderBatch.h"
+#include "cider/batch/CiderBatchUtils.h"
+#include "exec/module/batch/ArrowABI.h"
+
+class ArrowArrayBuilder {
+ public:
+  ArrowArrayBuilder() : row_num_(0), is_row_num_set_(false) {
+    array_ = CiderBatchUtils::allocateArrowArray();
+    schema_ = CiderBatchUtils::allocateArrowSchema();
+
+    schema_->format = "+s";
+    schema_->dictionary = nullptr;
+    schema_->release = CiderBatchUtils::ciderEmptyArrowSchemaReleaser;
+    // TODO: release stuff
+
+    array_->buffers = nullptr;
+    array_->n_buffers = 0;
+    array_->length = 0;
+    array_->offset = 0;
+    array_->release = CiderBatchUtils::ciderEmptyArrowArrayReleaser;
+  }
+
+  ArrowArrayBuilder& setTableName(const std::string& table_name) {
+    table_name_ = table_name;
+    schema_->name = table_name.c_str();
+    return *this;
+  }
+
+  ArrowArrayBuilder& setRowNum(int row_num) {
+    if (is_row_num_set_) {  // have set before, throw exception
+      CIDER_THROW(CiderCompileException, "row num have been set!");
+    }
+    is_row_num_set_ = true;
+    row_num_ = row_num;
+    return *this;
+  }
+
+  template <class T>
+  ArrowArrayBuilder& addColumn(const std::string& col_name,
+                               const ::substrait::Type& col_type,
+                               const std::vector<T>& col_data,
+                               const std::vector<bool>& null_data = {}) {
+    if (!is_row_num_set_ ||  // have not set row num, use this col_data's row num
+        row_num_ == 0) {     // previous columns are all empty
+      is_row_num_set_ = true;
+      row_num_ = col_data.size();
+    }
+    ArrowArray* current_array = new ArrowArray();
+    ArrowSchema* current_schema = new ArrowSchema();
+
+    current_schema->name = col_name.c_str();
+    current_schema->format = CiderBatchUtils::convertSubstraitTypeToArrowType(col_type);
+    current_schema->n_children = 0;
+    current_schema->children = nullptr;
+    current_schema->release = CiderBatchUtils::ciderEmptyArrowSchemaReleaser;
+
+    if (col_data.empty()) {
+      // append an empty buffer.
+      array_list_.push_back(nullptr);
+      schema_list_.push_back(current_schema);
+      return *this;
+    } else {
+      // check row num
+      if (row_num_ != col_data.size()) {
+        CIDER_THROW(CiderCompileException, "Row num is not equal to previous columns!");
+      }
+      CHECK_EQ(row_num_, col_data.size());
+      // check null data num
+      if (!null_data.empty()) {
+        CHECK_EQ(row_num_, null_data.size());
+      }
+
+      current_array->length = row_num_;
+      current_array->n_children = 0;
+      current_array->offset = 0;
+      current_array->buffers = (const void**)std::malloc(sizeof(void*) * 2);
+
+      void* null_buf = std::malloc((row_num_ % 8) ? (row_num_ / 8 + 1) : (row_num_ / 8));
```

Comment:
this would be more clear?
```
size_t null_size = (row_num_ + 7) >> 3;
void* null_buf = std::malloc(null_size);
std::memset(null_buf, 0xFF, null_size);
```